hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncBufferedMutatorImpl.java

```diff
@@ -153,24 +208,61 @@ Stream.<CompletableFuture<Void>> generate(CompletableFuture::new).limit(mutation
       bufferedSize += heapSize;
       if (bufferedSize >= writeBufferSize) {
         LOG.trace("Flushing because write buffer size {} reached", writeBufferSize);
-        internalFlush();
+        // drain now and send after releasing the lock
+        batch = drainBatch();
       } else if (maxMutations > 0 && this.mutations.size() >= maxMutations) {
         LOG.trace("Flushing because max mutations {} reached", maxMutations);
-        internalFlush();
+        batch = drainBatch();
       }
+    } finally {
+      lock.unlock();
+    }
+    // Send outside of lock
+    if (batch != null) {
+      sendBatch(batch);
     }
     return futures;
   }
 
   @Override
-  public synchronized void flush() {
-    internalFlush();
+  public void flush() {
```

Comment:
Considering that the only difference between `flush` and `close` is the one statement `closed = true`... is it worth creating a separate method that can be shared between the two with a parameter for whether to close?